sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctAttributesVisitor.scala

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeSet, ExpressionSet, NamedExpression}
+import org.apache.spark.sql.catalyst.plans.{LeftAnti, LeftSemi}
+
+/**
+ * A visitor pattern for traversing a [[LogicalPlan]] tree and propagate the distinct attributes.
+ */
+object DistinctAttributesVisitor extends LogicalPlanVisitor[Seq[AttributeSet]] {
+
+  private def allDistinctAttributes(
+      expressions: Seq[NamedExpression],
+      distinctAttributes: Seq[ExpressionSet]): Seq[AttributeSet] = {
+    expressions.toSet.subsets(distinctAttributes.map(_.size).min).filter { s =>
+      val references = s.map {
+        case a: Alias => a.child
+        case ne => ne
+      }
+      distinctAttributes.exists(_.equals(ExpressionSet(references)))
+    }.map(s => AttributeSet(s.map(_.toAttribute))).toSeq
+  }
+
+  override def default(p: LogicalPlan): Seq[AttributeSet] = {
+    Seq.empty[AttributeSet]
+  }
+
+  override def visitAggregate(p: Aggregate): Seq[AttributeSet] = {
+    val groupingExps = ExpressionSet(p.groupingExpressions) // handle group by a, a
```

Comment:
I think we should propagate the distinct keys from the child as well. This should be done in other places as well so we need to add a method for it, e.g.
```
 def addDistinctKey(keys: Set[AttributeSet], newExpressionSet: ExpressionSet): Set[AttributeSet]
```
The idea is: if `keys` already indicate the `newExpressionSet`, e.g. we have `[a, b]` in `keys`, we can ignore `newExpressionSet` if it's `[a, b, c]`. Else, we should clean up `keys` and add `newExpressionSet`. e.g. we have `[a, b, c]` in `keys` and the `newExpressionSet` is `[a, b]`, then we should remove `[a, b, c]` from `keys`.

Reply:
finally we can simply write `val distinctKeys = addDistinctKey(p.child.distinctKeys, ExpressionSet(p.groupingExpressions))` here.

Reply:
Does the distinct attributes related to the child distinct attributes? For example:
```sql
create table t(a int, b int, c int, d int, e int) using parquet;
select a, b, c, sum(d) from (select distinct * from t) t1 group by a, b, c;
```

Reply:
Does the distinct attributes related to the child of distinct attributes? For example:
```sql
create table t(a int, b int, c int, d int, e int) using parquet;
select a, b, c, sum(d) from (select distinct * from t) t1 group by a, b, c;
```